Deferred/Deferred.swift

```diff
@@ -6,39 +6,131 @@
+            // UnsafeMutablePointer.destroy() is faster than destroy(_:)
+            elementsPtr.destroy()
+        }
+    }
+    
+    func initializeWith(value: Value?) {
+        let box = value.map(Storage.init)
+        withUnsafeMutablePointerToElements { elementsPtr in
+            elementsPtr.initialize(box)
+        }
+    }
+    
+    func withValue(body: Value -> Void) {
+        withUnsafeMutablePointerToElements { elementsPtr in
+            guard let box = elementsPtr.memory else { return }
+            body(box.value)
+        }
+    }
+    
+    func fill(value: Value, onFill: SideData -> Void) -> Bool {
+        let box = Storage(value)
+        return withUnsafeMutablePointers { (valuePtr, elementsPtr) in
+            guard compareAndSwap(old: nil, new: box, to: elementsPtr) else { return false }
+            onFill(valuePtr.memory)
+            return true
+        }
+    }
+    
+}
+
+// A generic catch-all queue for when you just want to throw some work into the
+// concurrent pile. As an alternative to the `QOS_CLASS_UTILITY` global queue,
+// work dispatched onto this queue matches the QoS of the caller, which is
+// generally the right behavior.
+//
+// The technique is described and used in Core Foundation:
+// http://opensource.apple.com/source/CF/CF-1153.18/CFInternal.h
+private var genericQueue: dispatch_queue_t! {
+    return dispatch_get_global_queue(qos_class_self(), 0)
 }
 
 /// A deferred is a value that may become determined (or "filled") at some point
 /// in the future. Once a deferred value is determined, it cannot change.
 public struct Deferred<Value> {
-    typealias UponBlock = (dispatch_queue_t, Value -> ())
-    private typealias Protected = (protected: Value?, uponBlocks: [UponBlock])
     
-    private var protected: LockProtected<Protected>
+    private var storage = DeferredBuffer<Value, dispatch_block_t>.create {
+        dispatch_block_create(DISPATCH_BLOCK_NO_QOS_CLASS, {})
+    }
+
+    private func upon(queue: dispatch_queue_t, options inFlags: dispatch_block_flags_t, body: Value -> Void) -> dispatch_block_t {
+        let flags = dispatch_block_flags_t(inFlags.rawValue | DISPATCH_BLOCK_ASSIGN_CURRENT.rawValue)
+        let block = dispatch_block_create(flags) {
+            self.storage.withValue(body)
+        }
+        dispatch_block_notify(storage.value, queue, block)
+        return block
+    }
     
-    private init(_ value: Value?) {
-        protected = LockProtected(item: (value, []))
+    private func markFilled(onFilled: dispatch_block_t) {
+        dispatch_block_cancel(onFilled)
+        onFilled()
     }
+
+    // MARK: -
     
     /// Initialize an unfilled Deferred.
     public init() {
-        self.init(nil)
+        storage.initializeWith(nil)
     }
     
     /// Initialize a filled Deferred with the given value.
     public init(value: Value) {
-        self.init(value)
+        storage.initializeWith(value)
+        markFilled(storage.value)
```

Comment:
I think what these two lines are making more clear to me about my resistance to these changes is that as I read the code, while I trust that it is correct, it is not remotely close to what I would have imagined it would look like.  This doesn't make it wrong, in fact, it might be very correct, it just makes it really tough to maintain as it is not obvious.

A possible suggestion would be more documentation as to the high level approach that you are taking, and why it is the best approach.